broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/NewPartitionTransitionImpl.java

```diff
@@ -71,67 +73,48 @@ public void updateTransitionContext(final PartitionTransitionContext transitionC
     steps.forEach(step -> step.onNewRaftRole(context, role));
 
     final ActorFuture<Void> nextTransitionFuture = concurrencyControl.createFuture();
+    final var nextTransition =
+        new PartitionTransitionProcess(steps, concurrencyControl, context, term, role);
+    nextTransitionFuture.onComplete((v, e) -> lastTransition = nextTransition);
+
+    final var ongoingTransitionFuture =
+        currentTransitionFuture == null
+            ? concurrencyControl.createCompletedFuture()
+            : currentTransitionFuture;
+
+    // For safety reasons we have to immediately replace the current transition future with the next
+    // transition future, such that we make sure that we enqueue all transitions after another.
+    //
+    // This means we will always add the next transition to the tail, this can become a chain of
+    // futures if we have many transitions after another. Ideally we would execute after the current
+    // transition only the last enqueue one. This should be implemented as soon we have time for it.
+    currentTransitionFuture = nextTransitionFuture;
```

Comment:
This is not thread safe

Reply:
Doesn't need to be; it is called in the same actor

Reply:
It is enqueue here https://github.com/camunda-cloud/zeebe/blob/develop/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartition.java#L105

I removed the additional #run call on purpose, since this was unnecessary scheduling. In my opinion we need the guarantee that we always run in the same actor/thread in these partitions transitions. We expect that also on other places.

Reply:
That is just an implementation detail and can change any time. I am not comfortable implementing classes that make assumptions in how they are being used. If you insist, then at lest add a comment and a runtime check that we are indeed in an Actor thread.

Reply:
In general I agree with you that the we should not expect how it is used, but this is currently anyway the case. Otherwise you have to call actor.run on all places right? At some point you have to make guarantees and expectation I think. In the processor for example we know and expect everything is run in one actor. Or in the ExporterDirector etc. 

The runtime check might be problematic for the test where we are running with the other concurrency Control.

I can add the run again, but would like to hear @deepthidevaki or @npepinpe opinion

Reply:
I think this goes a bit in what we discussed this morning. I would like a clear separation between async (i.e. actor behaviors) and sync (e.g. the partition transition process) classes, so that it's very clear when one is called and one isn't.

It makes sense for performance reason and semantic to not schedule every single method, but it should be immediately obvious when something is meant to be used outside of its actor context or not. I haven't had time yet to dig into this and produce a template, but if someone has an idea or wants time for that I'm happy to discuss it.

Anyway, going back to the problem at hand - I haven't had a deep look into the new transition behavior, so it's hard for me to gauge the risk of this particular method being called from a different thread context. If we expect the `NewPartitionTransitionImpl` to deal with asynchronicity, then it makes sense to have something, whether at compile time or runtime, to check ensure we call the method from the correct context. If we don't, but the fear is that someone might do so in the future, then I would hope documentation (whether javadoc, proper naming, or a combination) would be enough to dissuade people, and highlight the issue possibly to future reviewers.

I guess this isn't as helpful as you'd hoped :smile: Let me know if I need to take a deeper look to make a final decision, otherwise I trust you to figure something out :slightly_smiling_face:

Reply:
It might be safe to add run, then it is not required to execute the transition on ZeebePartition actor. Right now it does not make any difference. But since this class is self contained it make sense to not make the assumption that it will be run on the actor which calls this method.